src/commands/mobile_app/upload.rs

```diff
@@ -102,10 +103,53 @@ pub fn execute(matches: &ArgMatches) -> Result<()> {
         .map(Cow::Borrowed)
         .or_else(|| vcs::find_head().ok().map(Cow::Owned));
 
-    let base_sha = matches.get_one("base_sha").map(String::as_str);
-    let vcs_provider = matches.get_one("vcs_provider").map(String::as_str);
-    let head_repo_name = matches.get_one("head_repo_name").map(String::as_str);
+    let cached_remote = config.get_cached_vcs_remote();
+    // Try to open the git repository and find the remote, but handle errors gracefully.
+    let (vcs_provider, head_repo_name) = {
+        // Try to open the repo and get the remote URL, but don't fail if not in a repo.
+        let remote_url = match git2::Repository::open_from_env().and_then(|repo| {
+            repo.find_remote(&cached_remote).and_then(|remote| {
+                remote
+                    .url()
+                    .map(|url| url.to_owned())
+                    .ok_or_else(|| git2::Error::from_str("No remote URL found"))
+            })
+        }) {
```

Comment:
This reads much better if you extract it into a function (say, `git_repo_remote_url`) because then you can use `?` and don't have to cobble everything together with `and_then`. The lack of `try` blocks really is a pain sometimes.
```rust
fn git_repo_remote_url(cached_remote: &str) -> Result<String, git2::Error> {
    let repo = git2::Repository::open_from_env()?;
    let remote = repo.find_remote(cached_remote)?;
    remote
        .url()
        .map(|url| url.to_owned())
        .ok_or_else(|| git2::Error::from_str("No remote URL found"))
}
```

```suggestion
        let remote_url = match git_repo_remote_url(&cached_remote) {
```